lib/vrl/compiler/src/expression.rs

```diff
@@ -69,7 +69,7 @@ pub trait Expression: Send + Sync + fmt::Debug + DynClone {
 
     /// Updates the state if necessary.
     /// By default it does nothing.
-    fn update_state(&self, _state: &mut crate::State) -> Result<(), ExpressionError> {
+    fn update_state(&mut self, _state: &mut crate::State) -> Result<(), ExpressionError> {
```

Comment:
I'd prefer it if we could keep this as an immutable reference, to prevent compiled function state from shifting under our feet as we do compile-time checks.

I'm not convinced we can find a way to do so, but it would be worth taking a crack at it.

Reply:
When the vrl function adds an index it is passed an index handle. This is then passed to the search so the search can instantly retrieve the index to use in the search. This needs to be mutable so that the function can store that handle.

It's not strictly necessary, but it saves an allocation and a little bit of time searching for the index to use on lookup. It would be worth benchmarking because maybe it doesn't save any real time at all, I only put this in when I was running dodgy benchmarking code and was doing everything I could to get the time down.

Reply:
Got it. Let's keep this unresolved for now then, and circle back to it once we have proper benchmarks to compare the differences.

Reply:
I agree that we should model the interaction clearly. If it doesn't need a mutable reference, but we are just doing so for performance reasons, it'd be good to have some benchmarks to back that up and then some comments in the code describing why the decision was made.